tslearn/early_classification.py

```diff
@@ -0,0 +1,269 @@
+            self.classifiers_[t].fit(X1[:, :t], y1)
+            for k in range(0, self.n_clusters):
+                index = (c_k2 == k)
+                if index.shape[0] != 0:
+                    X2_current_cluster = X2[index, :t]
+                    y2_current_cluster = y2[index]
+                    y2_hat = self.classifiers_[t].predict(
+                        X2_current_cluster[:, :t]
+                    )
+                    conf_matrix = confusion_matrix(
+                        y2_current_cluster, y2_hat, labels=y_classes_indices, normalize="true"
+                    )
+                    #The diagonal of the confusion matrices must be put to 0
+                    # to fullfil the criterions of the cost function
+                    np.fill_diagonal(conf_matrix, 0)
+                    self.pyhatyck_[t - self.minimum_time_stamp, k] = conf_matrix
+
+    def get_cluster_probas(self, X):
+        """
+        This function computes the probabilities of the time series xt to be in a cluster of the model
+
+        Parameters
+        ----------
+        X: vector
+            a time series that may be truncated
+
+        Returns
+        -------
+        vector : the probabilities of the given time series to be in the clusters of the model
+        """
+        diffs = X[np.newaxis, :] - self.cluster_.cluster_centers_[:, :len(X)]
+        distances_clusters = np.linalg.norm(diffs, axis=(1, 2))
+        minimum_distance = np.min(distances_clusters)
+        average_distance = np.mean(distances_clusters)
+        minimum_distance_to_average = (
+            average_distance - minimum_distance
+        ) / average_distance
+        delta = (average_distance - distances_clusters) / average_distance
+        probas = 1.0 / (
+            np.exp(self.lamb * minimum_distance_to_average)
+            + np.exp(self.lamb * (minimum_distance_to_average - delta))
+            + 1e-6
+        )
+        return probas / probas.sum()
+
+    def _expected_cost(self, X, tau):
+        """
+        From a incomplete series xt, this function compute the expected cost of a prediction made at time "last time of
+        xt + tau"
+
+        Parameters
+        ----------
+        X: vector
+            a time series that may be truncated for which we want to have the cost
+        tau: int
+            gives the future time for which the cost is calculated
+
+        Returns
+        --------
+        float : the computed cost
+        """
+        proba_clusters = self.get_cluster_probas(X=X)
+        truncated_t = X.shape[0]
+        sum_pyhatyck = np.sum(self.pyhatyck_[truncated_t + tau - self.minimum_time_stamp], axis=1)
+        sum_pyhatyck = np.transpose(sum_pyhatyck)
+        sum_global = np.sum((sum_pyhatyck * self.pyck_), axis=0)
+        cost = np.dot(proba_clusters, sum_global)
+        return cost + self._cost_time(truncated_t + tau)
+
+    def cost_function_minimizer(self, end_of_time, xt):
```

Comment:
One thing I just noticed is that in this code there are several different notation for similar objects. Typically, I would tend to use `Xi` to denote a single time series in all methods, so that going through the code is easier. And use `X` only when this is a dataset (not when it is a single series), which would require to change the variable name in `get_cluster_proba` and in `_expected_cost`.